gdal/ogr/ogrsf_frmts/dgn/ogrdgnlayer.cpp

```diff
@@ -329,13 +364,67 @@ OGRFeature *OGRDGNLayer::ElementToFeature( DGNElemCore *psElement, int nRecLevel
     int anMSLink[MAX_LINK];
     anMSLink[0] = 0;
 
+    CPLJSONObject uLinkData;
+    CPLJSONArray previousValues;
+
     int iLink = 0;
     int nLinkCount = 0;
+    int uLinkCount = 0;
+
+    int nLinkType = 0;
+    int nLinkSize = 0;
+
+    unsigned char *pabyData = DGNGetLinkage( hDGN, psElement, iLink, &nLinkType,
+                              anEntityNum + iLink, anMSLink + iLink, &nLinkSize);
 
-    unsigned char *pabyData = DGNGetLinkage( hDGN, psElement, iLink, nullptr,
-                              anEntityNum + iLink, anMSLink + iLink, nullptr );
     while( pabyData )
     {
+        previousValues = uLinkData.GetArray( std::to_string(nLinkType) );
+        if (!previousValues.IsValid() ) 
+        {
+            uLinkData.Add( std::to_string(nLinkType), CPLJSONArray() );
+            previousValues = uLinkData.GetArray( std::to_string(nLinkType) );
+        } 
+        char *pszAsHex = CPLBEBinaryToHex( nLinkSize - 4, pabyData + 4 );
```

Comment:
why not using plain CPLBinaryToHex() ? It seems to me more standard to report the 4 high-order bits as the first hexadecimal character

Reply:
Agree. 

The order needs to be reverted, just for the data related with `application id = 0x1995`. For this application id, I know that the data order has to be reversed.

Reply:
> The order needs to be reverted, just for the data related with `application id = 0x1995`. For this application id, I know that the data order has to be reversed.

That doesn't make sense to me why you'd need to do that. It is also a bit odd to call something big-endian whereas we are talking about the ordering of the high and low 4 bits of a byte, whereas big-endian normally apply to multi-byte integers/float. If that function is really needed, I'd suggest to do just add a logic to permute bytes after calling CPLBinaryToHex() instead of adding something in the CPL namespace

Reply:
@rouault I've checked data for other application ids again. It seems to me that we need to reverse the byte order for other application IDs, as well. I'm not sure if we have to do this for all application IDs, so this comment might be _useful for future reference_. We return the data as raw, but people might need to swap byte order.

Example using CAD file used on [ticket 1387](https://trac.osgeo.org/gdal/ticket/1387) 

For DB linkages, for example, `application id = 24721`, we have:

Using ODA's API:
```
{ "24721": [ { "tableId": 33, "MSLink": 2406, "type": "Oracle" } ] }
```
Using DGN driver, with raw hex, without reverse byte order:
```
{ "24721": [ { "raw": "810F21006409000000000000", "type": "Oracle", "size": 16 } ] }
```
Reversing the byte order, we have:
<pre>
{ "24721": [ { "raw": "0F81 <b>0021 </b><b>0964</b> 000000000000", "type": "Oracle", "size": 16 } ] }
</pre>
With the reversed byte order, we can find the `tableId: 0x0021`  (decimal 33) and `MSLink: 0x0964` (decimal 2406). The reversed order matches ODA's API behaviour.

What do you think?

Reply:
oh.... I misunderstood what CPLBEBinaryToHex() do. I thought it switched the upper and lower bits of each byte. But it actually deals with 16-bit words in big-endian order.  But I don't see the need for that function though . This is more a documentation issue. If it is documented that 2100 should be interpreted as 0x21 | (0x00 << 8) and 6409 as 0x64 | (0x09 << 8), then things are OK

Reply:
My fault. I should have been more verbose. The `BE` in `CPLBEBinaryToHex` stands for Big Endian, but it is too short.

While little/big endian are related to CPU architectures, the same concept in strings is less obvious. That's why I added this `CPLBEBinaryToHex()`. But I don't have a strong opinion. If you prefer to dump it as it is, we can do it that way. I would prefer to have the output (the hex string) properly ordered.

Please let me know what do you prefer @rouault (and thank you for all the feedback).

Reply:
When dumping raw content, I find it better that it is dump as it, without doing byte swapping. If the raw content is always a pair of 16-bit words (haven't checked what DGN says about that), then "raw" could be an array of decoded integers instead. If the interpretation of the content depends on the user link, then raw is fine.

Reply:
Thank you for the feedback @rouault 

I was using this [reference](http://www.la-solutions.co.uk/content/Databases/ElementData.htm):

* **A user data linkage is stored in an array of short (16-bit) words. Each array contains an application ID, an attribute ID, and a word count, followed by an arbitrary amount of linkage data.**
* **It's up to you to figure out how to encode and decode your user data. You probably want to write one function that wraps raw data with your Application ID, and another that can extract the Application ID and raw data from a data linkage.**

Definitely, the interpretation of the content depends on the user link.

If you don't mind, I'll follow different approaches for the DGN driver and for the DGNV8.

### DGN driver
Dump the content as it is, without any interpretation or byte swapping. 

### DGNV8 driver
Dump the contents with interpretation, since we can use ODA API to get structured data for known application ids. There are known application ids like `OdDgDBLinkage` that we can take advantage. For the unknown ones, we still have to dump the contents as it is.

### Documentation
I'll update the driver's documentation to reflect this.

Sound reasonable to you?

Reply:
sounds good